ddsrouter_event/include/ddsrouter_event/wait/impl/WaitHandler.ipp

```diff
@@ -0,0 +1,196 @@
+
+/**
+ * @file WaitHandler.ipp
+ */
+
+#include <ddsrouter_utils/Log.hpp>
+#include <ddsrouter_utils/Time.hpp>
+
+#ifndef _DDSROUTEREVENT_IMPL_WAITHANDLER_IPP_
+#define _DDSROUTEREVENT_IMPL_WAITHANDLER_IPP_
+
+namespace eprosima {
+namespace ddsrouter {
+namespace event {
+
+template <typename T>
+WaitHandler<T>::WaitHandler(
+        bool enabled /* = true */)
+    : enabled_(enabled)
+    , threads_waiting_(0)
+{
+}
+
+template <typename T>
+WaitHandler<T>::WaitHandler(
+        T init_value,
+        bool enabled /* = true */)
+    : enabled_(enabled)
+    , threads_waiting_(0)
+    , value_(init_value)
+{
+}
+
+template <typename T>
+WaitHandler<T>::~WaitHandler()
+{
+    blocking_disable();
+}
+
+template <typename T>
+void WaitHandler<T>::enable() noexcept
+{
+    // If disable, enable it. Otherwise do nothing
+    if(!enabled_.load())
+    {
+        // WARNING: enabled_ should be modified with mutex taken
+        std::lock_guard<std::mutex> lock(wait_condition_variable_mutex_);
+        logDebug(DDSROUTER_WAIT, "Enabling WaitHandler.");
+        enabled_.store(true);
+    }
+    else
+    {
+        logDebug(DDSROUTER_WAIT, "Enabling not disabled WaitHandler.");
+    }
+}
+
+template <typename T>
+void WaitHandler<T>::disable() noexcept
+{
+    // If enable, disable it. Otherwise do nothing
+    if(enabled_.load())
+    {
+        // WARNING: enabled_ should be modified with mutex taken
+        std::lock_guard<std::mutex> lock(wait_condition_variable_mutex_);
+        logDebug(DDSROUTER_WAIT, "Disabling WaitHandler.");
+        enabled_.store(false);
+
+        // Do not block for awaken
+        wait_condition_variable_.notify_all();
+    }
+    else
+    {
+        logDebug(DDSROUTER_WAIT, "Disabling not enabled WaitHandler.");
+    }
+}
+
+template <typename T>
+void WaitHandler<T>::blocking_disable() noexcept
+{
+    // Disable this object
+    disable();
+
+    // Wait till every thread has finished
+    while(threads_waiting_.load() > 0)
+    {
+        // Lock so this thread has no priority over closin threads
+        {
+            std::lock_guard<std::mutex> lock(wait_condition_variable_mutex_);
+        }
+
+        // Notify in case there are still threads waiting to be awaken
+        wait_condition_variable_.notify_all();
+    }
+}
+
+template <typename T>
+bool WaitHandler<T>::enabled() const noexcept
+{
+    return enabled_.load();
+}
+
+template <typename T>
+AwakeReason WaitHandler<T>::wait(
+        std::function<bool(const T&)> predicate,
+        const utils::Duration_ms& timeout /* = 0 */)
+{
+    // Do wait with mutex taken
+    std::unique_lock<std::mutex> lock(wait_condition_variable_mutex_);
+
+    // Check if it is disabled and exit
+    if (!enabled())
+    {
+        return AwakeReason::DISABLED;
+    }
+
+    // Increment number of threads waiting
+    // WARNING: mutex must be taken
+    threads_waiting_++;
+
+    bool finished_for_timeout = false;
+    utils::Timestamp time_to_wait_until;
```

Comment:
```suggestion
    utils::Timestamp time_to_wait;
```

Reply:
I think is better this name.
`time_to_wait` could be misunderstood for time that should be waiting, but this is not the case.
The case is the time until the cv must wait.